File: native/swift/Sources/BlocksCodegen/CodegenModel.swift
import Foundation

// MARK: - Resolved Codegen Model

/// A fully resolved, language-independent type ready for translation to Swift.
indirect enum ResolvedType {
    case primitive(PrimitiveKind, constraints: Constraints = .empty)
    /// A type whose Swift representation is determined by the JSON Schema
    /// `format` keyword (e.g. `format: "date-time"` → `Date`).
    case formattedType(FormatKind, constraints: Constraints = .empty)
    /// Named record. When `additionalPropertiesType` is non-nil the record is
    /// "open" — the spec declared `T & Record<string, V>`. Codegen emits an
    /// `attributes: [String: V]` field and overrides Codable so that map
    /// flattens at the JSON top level. When `embeddedUnion` is set, the
    /// record carries an inner discriminated union flattened at the same
    /// JSON envelope (one `oneOf` arm holds outer fields + nested oneOf —
    /// see Cognito's regrouped `confirmSignIn` arm).
    case record(name: String, fields: [ResolvedField], additionalPropertiesType: ResolvedType?, embeddedUnion: ResolvedType?)
    case `enum`(name: String, values: [String])
    case list(elementType: ResolvedType, constraints: Constraints = .empty)
    case nullable(inner: ResolvedType)
    case union(name: String, variants: [UnionVariant], discriminator: DiscriminatorInfo?)
    case typeReference(name: String)
    case transferable(blocksType: String, typeArgs: [ResolvedType])
    case map(valueType: ResolvedType)
}

/// Format keyword mappings that produce a non-`String` Swift type.
enum FormatKind {
    case dateTime  // → Foundation.Date
    case date      // → Foundation.Date (via ISO 8601 day-only)
    case time      // → String (Foundation has no time-only type)
    case uuid      // → Foundation.UUID
    case uri       // → Foundation.URL
}

/// JSON Schema constraint annotations carried alongside a primitive / list type.
/// Code generation may emit `precondition` checks at construct time when these
/// are set; today they are decoded but only the `format` field is consumed
/// (mapped into `ResolvedType.formattedType` upstream).
struct Constraints: Equatable {
    let format: String?
    let minLength: Int?
    let maxLength: Int?
    let pattern: String?
    let minimum: Double?
    let maximum: Double?
    let exclusiveMinimum: Double?
    let exclusiveMaximum: Double?
    let multipleOf: Double?
    let minItems: Int?
    let maxItems: Int?

    init(
        format: String? = nil,
        minLength: Int? = nil,
        maxLength: Int? = nil,
        pattern: String? = nil,
        minimum: Double? = nil,
        maximum: Double? = nil,
        exclusiveMinimum: Double? = nil,
        exclusiveMaximum: Double? = nil,
        multipleOf: Double? = nil,
        minItems: Int? = nil,
        maxItems: Int? = nil
    ) {
        self.format = format
        self.minLength = minLength
        self.maxLength = maxLength
        self.pattern = pattern
        self.minimum = minimum
        self.maximum = maximum
        self.exclusiveMinimum = exclusiveMinimum
        self.exclusiveMaximum = exclusiveMaximum
        self.multipleOf = multipleOf
        self.minItems = minItems
        self.maxItems = maxItems
    }

    static let empty = Constraints()

    var isEmpty: Bool { self == Constraints.empty }
}

struct ResolvedField {
    let name: String
    let type: ResolvedType
    let required: Bool
    let description: String?
    /// `default` value from the spec, encoded as raw JSON. Emitted as a Swift
    /// literal default when present (string / number / boolean / null only).
    let defaultValue: String?

    init(name: String, type: ResolvedType, required: Bool, description: String? = nil, defaultValue: String? = nil) {
        self.name = name
        self.type = type
        self.required = required
        self.description = description
        self.defaultValue = defaultValue
    }
}

struct UnionVariant {
    let name: String
    let fields: [ResolvedField]
    let discriminatorValue: String?
    /// Set when the variant payload is an existing named type (e.g. another
    /// union or a typeReference) rather than a fresh inline record. Codegen
    /// uses this to emit `case foo(ExistingType)` instead of synthesizing
    /// a new struct with `fields`.
    let payloadTypeName: String?
    /// Open-shape value type (e.g. `String` for `Record<string, string>`)
    /// when the variant's spec entry carried `additionalProperties`.
    /// Propagates so the synthesized variant struct gets the same
    /// `attributes: [String: V]` treatment as a top-level open record.
    let additionalPropertiesType: ResolvedType?
    /// Inner discriminated alternative whose fields flatten into the same
    /// JSON envelope (regrouped union arm). Codegen emits a merged Codable
    /// for the synthesized variant struct.
    let embeddedUnion: ResolvedType?
    /// Types declared inside this variant's fields that should be scoped
    /// under the variant struct (e.g. `NextStep` inside `IsSignedInFalse`).
    let nestedTypes: [NestedTypeNode]

    init(name: String, fields: [ResolvedField], discriminatorValue: String? = nil, payloadTypeName: String? = nil, additionalPropertiesType: ResolvedType? = nil, embeddedUnion: ResolvedType? = nil, nestedTypes: [NestedTypeNode] = []) {
        self.name = name
        self.fields = fields
        self.discriminatorValue = discriminatorValue
        self.payloadTypeName = payloadTypeName
        self.additionalPropertiesType = additionalPropertiesType
        self.embeddedUnion = embeddedUnion
        self.nestedTypes = nestedTypes
    }
}

struct DiscriminatorInfo {
    let fieldName: String
    let variants: [String: String] // discriminator value → variant name
}

// MARK: - Operations

struct OperationParameter {
    let name: String
    let type: ResolvedType
    let required: Bool
    let description: String?

    init(name: String, type: ResolvedType, required: Bool, description: String? = nil) {
        self.name = name
        self.type = type
        self.required = required
        self.description = description
    }
}

struct OperationResult {
    let type: ResolvedType
    let description: String?

    init(type: ResolvedType, description: String? = nil) {
        self.type = type
        self.description = description
    }
}

struct Operation {
    let name: String
    let parameters: [OperationParameter]
    let result: OperationResult
    let description: String?
    let nestedTypes: [NestedTypeNode]

    init(name: String, parameters: [OperationParameter], result: OperationResult, description: String? = nil, nestedTypes: [NestedTypeNode] = []) {
        self.name = name
        self.parameters = parameters
        self.result = result
        self.description = description
        self.nestedTypes = nestedTypes
    }
}

struct NestedTypeNode {
    let name: String
    let type: ResolvedType
    let children: [NestedTypeNode]

    init(name: String, type: ResolvedType, children: [NestedTypeNode] = []) {
        self.name = name
        self.type = type
        self.children = children
    }
}

struct APINamespace {
    let name: String
    let operations: [Operation]
}

// MARK: - Type Definitions

struct TypeDefinition {
    let name: String
    let type: ResolvedType
    let shortName: String

    init(name: String, type: ResolvedType, shortName: String? = nil) {
        self.name = name
        self.type = type
        self.shortName = shortName ?? name
    }
}

// MARK: - Root Model

public struct CodegenModel {
    let apiNamespaces: [APINamespace]
    let typeDefinitions: [TypeDefinition]
    let servers: [ServerDefinition]
}

File: native/kotlin/codegen/src/main/java/com/aws/blocks/kotlin/model/CodegenModel.kt
package com.aws.blocks.kotlin.model

import kotlinx.serialization.json.JsonElement

// ── Primitive Kind ────────────────────────────────────────────────

enum class PrimitiveKind {
    STRING, BOOLEAN, INTEGER, NUMBER, VOID, UNKNOWN
}

// ── Format Kind ──────────────────────────────────────────────────

enum class FormatKind {
    DATE_TIME,  // kotlinx.datetime.Instant
    DATE,       // kotlinx.datetime.LocalDate
    TIME,       // kotlinx.datetime.LocalTime
    UUID,       // kotlin.uuid.Uuid
}

// ── ResolvedType Sealed Hierarchy ─────────────────────────────────

/** A fully resolved, language-independent type ready for translation to any target language. */
sealed interface ResolvedType {

    /** Primitive types: String, Boolean, Int, Double, Unit, Any */
    data class Primitive(val kind: PrimitiveKind, val constraints: Constraints = Constraints.EMPTY) : ResolvedType

    /** A type whose Kotlin representation is determined by its JSON Schema `format` keyword. */
    data class FormattedType(val format: FormatKind, val constraints: Constraints = Constraints.EMPTY) : ResolvedType

    /** Named record type with resolved fields. */
    data class Record(
        val name: String,
        val fields: List<ResolvedField>,
        val description: String? = null,
        val additionalPropertiesType: ResolvedType? = null,
    ) : ResolvedType

    /** Enum type with string literal values. */
    data class Enum(
        val name: String,
        val values: List<String>,
    ) : ResolvedType

    /** List/array type. */
    data class ListType(val elementType: ResolvedType, val constraints: Constraints = Constraints.EMPTY) : ResolvedType

    /** Map type (object with additionalProperties). */
    data class MapType(val valueType: ResolvedType) : ResolvedType

    /** Tuple type (fixed-length heterogeneous list). */
    data class TupleType(val name: String, val elements: List<ResolvedType>) : ResolvedType

    /** Nullable wrapper. */
    data class Nullable(val inner: ResolvedType) : ResolvedType

    /** Sealed/union type with variant members. */
    data class Union(
        val name: String,
        val variants: List<UnionVariant>,
        val discriminator: DiscriminatorInfo?,
    ) : ResolvedType

    /** Reference to a named type defined elsewhere in the model. */
    data class TypeReference(val name: String) : ResolvedType

    /** A transferable type that maps to a hydrated live object (e.g., RealtimeChannel, FileDownloadHandle). */
    data class Transferable(val transferableName: String, val typeArgs: List<ResolvedType>) : ResolvedType
}

// ── Supporting Types ──────────────────────────────────────────────

data class ResolvedField(
    val name: String,
    val type: ResolvedType,
    val required: Boolean,
    val description: String? = null,
    val constraints: Constraints = Constraints.EMPTY,
    val defaultValue: JsonElement? = null,
)

data class UnionVariant(
    val name: String,
    val fields: List<ResolvedField>,
    val discriminatorValue: String? = null,
    val embeddedUnion: ResolvedType.Union? = null,
    val payloadTypeName: String? = null,
    val additionalPropertiesType: ResolvedType? = null,
    val nestedTypes: List<NestedTypeNode> = emptyList(),
)

data class DiscriminatorInfo(
    val fieldName: String,
    /** Discriminator value → variant name */
    val variants: Map<String, String>,
    val type: DiscriminatorType = DiscriminatorType.STRING,
)

enum class DiscriminatorType {
    STRING,
    BOOLEAN,
}

data class ServerDefinition(
    val name: String,
    val url: String,
)

// ── Root Model and Structure ──────────────────────────────────────

data class CodegenModel(
    val apiNamespaces: List<ApiNamespace>,
    val typeDefinitions: List<TypeDefinition>,
    val servers: List<ServerDefinition>,
    val endpoint: String?,
)

data class ApiNamespace(
    val name: String,
    val operations: List<Operation>,
)

data class Operation(
    val name: String,
    val parameters: List<OperationParameter>,
    val result: OperationResult,
    val description: String?,
    val nestedTypes: List<NestedTypeNode> = emptyList(),
)

data class NestedTypeNode(
    val name: String,
    val type: ResolvedType,
    val children: List<NestedTypeNode> = emptyList(),
)

data class OperationParameter(
    val name: String,
    val type: ResolvedType,
    val required: Boolean,
    val description: String?,
)

data class OperationResult(
    val type: ResolvedType,
    val description: String?
)

data class TypeDefinition(
    val name: String,
    val type: ResolvedType,
    val shortName: String = name,
    val parentSchema: String? = null,
)

